aws/sdk/examples/using_native_tls_instead_of_rustls/src/main.rs

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+/// The SDK defaults to using RusTLS by default but you can also use [`native_tls`](https://github.com/sfackler/rust-native-tls)
+/// which will choose a TLS implementation appropriate for your platform. This example looks much like
+/// any other. Activating and deactivating `features` in your app's `Cargo.toml` is all that's needed.
+#[tokio::main]
+async fn main() -> Result<(), aws_sdk_s3::Error> {
+    tracing_subscriber::fmt::init();
+
+    let shared_config = aws_config::load_from_env().await;
+
+    let s3_config = aws_sdk_s3::Config::from(&shared_config);
+    let client = aws_sdk_s3::Client::from_conf(s3_config);
+
+    let resp = client.list_buckets().send().await?;
+
+    for bucket in resp.buckets().unwrap_or_default() {
+        println!("bucket: {:?}", bucket.name().unwrap_or_default())
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    /// You can run this test to ensure that this example is only using `native-tls`
+    /// and that nothing is pulling in `rustls` as a dependency
+    #[test]
+    #[should_panic = "error: package ID specification `rustls` did not match any packages"]
+    fn test_rustls_is_not_in_dependency_tree() {
+        let cargo_location = std::env::var("CARGO").unwrap();
+        let cargo_command = std::process::Command::new(&cargo_location)
+            .arg("tree")
+            .arg("--invert")
+            .arg("rustls")
+            .output()
```

Comment:
Clever!

Does this differentiate between something being wrong with the cargo binary (not able to execute or something) vs. exiting with a non-zero status?

Reply:
This only checks that Cargo writes a specific message to stderr when the command is run. The possible paths I see are:

1. rustls is not in tree, message is logged to stderr, we panic with that message and match on it with `should_panic`, it's a match so that's a test pass
2. some other failure occurs, message is logged to stderr, we panic with that message and match on it with `should_panic`, It's not a match so that's a test fail
3. rustls **is** in the tree, print tree and exit, test fails cause no panic

So, to answer your question: yes, it does differentiate